IMPORT INTO autodetect file format and wildcards for file path (#21197)
pingcap/docs: sql-statements/sql-statement-import-into.md

### fileLocation

It specifies where your data files are and which files to import. You can point to a single file or use wildcards to match many files.

> - Cloud storage (Amazon S3 or GCS): Provide the full object-storage URI, formatted as described in [URI Formats of External Storage Services](/external-storage-uri.md).

Reviewer: ```suggestion
- Cloud storage (Amazon S3 or GCS): provide the full object storage URI, formatted as described in [URI Formats of External Storage Services](/external-storage-uri.md).
```

### fileLocation

It specifies where your data files are and which files to import. You may point to a single file or use wildcards to match many files.

- Amazon S3 or GCS URI path: for URI configuration details, see [URI Formats of External Storage Services](/external-storage-uri.md).

> - TiDB local file path: it must be an absolute path. It is recommended to use a recognized extension (`.csv`, `.sql`, `.parquet`). If the file has no extension, TiDB treats it as CSV. Ensure the specified files exist on the TiDB node where your session is connected, and confirm you have the required `FILE` privilege.

Author: ```suggestion
- TiDB local file path: The path must be absolute. Ensure the specified path and files exist on the TiDB node where your session is connected, and confirm you have the required `FILE` privilege.
```

### fileLocation

It specifies where your data files are and which files to import. You can point to a single file or use wildcards to match many files.

> - Amazon S3 or GCS URI path: for URI configuration details, see [URI Formats of External Storage Services](/external-storage-uri.md).

Author: ```suggestion
- Cloud storage (Amazon S3 or GCS): Provide the full object-storage URI, formatted as described in [URI Formats of External Storage Services](/external-storage-uri.md).
```

- TiDB local file path: The path must be absolute. Ensure the specified path and files exist on the TiDB node where your session is connected, and confirm you have the required `FILE` privilege.

> **Note:**
>
> If [SEM](/system-variables.md#tidb_enable_enhanced_security) is enabled in the target cluster, the `fileLocation` cannot be specified as a local file path.

> In the `fileLocation` parameter, you can specify a single file or use wildcards (`*` and `[]`) to match multiple files. Wildcards can be used to match sub-path segments (e.g., a directory level) and filenames. For example, if your files are stored on Amazon S3, you can configure the parameter like this:

Author: Reverting this change, as directory-level wildcards are not yet supported by the implementation. This also resolves the `vale` linter error.

```suggestion
In the `fileLocation` parameter, you can specify a single file, or use the `*` and `[]` wildcards to match multiple files for import. Note that the wildcard can only be used in the file name, because it does not match directories or recursively match files in subdirectories. Taking files stored on Amazon S3 as examples, you can configure the parameter as follows:
```

- Import all files with the `.csv` suffix in a specified path: `s3://<bucket-name>/path/to/data/*.csv`
- Import all files with the `foo` prefix in a specified path: `s3://<bucket-name>/path/to/data/foo*`
- Import all files with the `foo` prefix and the `.csv` suffix in a specified path: `s3://<bucket-name>/path/to/data/foo*.csv`
- Import `1.csv` and `2.csv` in a specified path: `s3://<bucket-name>/path/to/data/[12].csv`. This is useful for importing a specific, non-sequential set of files.
- Import `1.csv`, `2.csv`, and `3.csv` using a range: `s3://<bucket-name>/path/to/data/[1-3].csv`

> 

Author: ```suggestion
- Import files with a single character name, except `1.csv` or `2.csv` using `^` for negation: `s3://<bucket-name>/path/to/data/[^12].csv`

```

- Import all files with the `foo` prefix in a specified path: `s3://<bucket-name>/path/to/data/foo*`
- Import all files with the `foo` prefix and the `.csv` suffix in a specified path: `s3://<bucket-name>/path/to/data/foo*.csv`
- Import `1.csv` and `2.csv` in a specified path: `s3://<bucket-name>/path/to/data/[12].csv`

> - Import `foo.csv` files from all immediate sub-paths: `s3://<bucket-name>/path/to/*/foo.csv` (add another `*/` for each extra directory level, for example `path/to/*/*/foo.csv`)

Author: Removing the incorrect directory-level wildcard example on line 131. 

```suggestion
```

- Import all files with the `foo` prefix and the `.csv` suffix in a specified path: `s3://<bucket-name>/path/to/data/foo*.csv`

> - Import `1.csv` and `2.csv` in a specified path: `s3://<bucket-name>/path/to/data/[12].csv`

Author: ```suggestion
- Import `1.csv` and `2.csv` in a specified path: `s3://<bucket-name>/path/to/data/[12].csv`. This is useful for importing a specific, non-sequential set of files.
- Import `1.csv`, `2.csv`, and `3.csv` using a range: `s3://<bucket-name>/path/to/data/[1-3].csv`
```

>
> - The Snappy compressed file must be in the [official Snappy format](https://github.com/google/snappy). Other variants of Snappy compression are not supported.
> - Because TiDB Lightning cannot concurrently decompress a single large compressed file, the size of the compressed file affects the import speed. It is recommended that a source file is no greater than 256 MiB after decompression.

> > - When `FORMAT` is omitted, TiDB first removes one compression suffix from the file name, then inspects the remaining extension to choose `CSV` or `SQL`.

Reviewer: ```suggestion
> - When the `FORMAT` clause is omitted, TiDB first removes one compression suffix from the file name, then inspects the remaining extension to determine whether the file is  `CSV` or `SQL`.
```

IMPORT INTO t FROM '/path/to/file-0[13].csv';
```

#### Import data from a nested directory structure

You can use wildcards to import data from a common directory structure, such as one organized by date. For example, assume your sales data is stored in S3 and organized by year and quarter:

```
/path/to/sales-data/2023/q1/data.csv
/path/to/sales-data/2023/q2/data.csv
/path/to/sales-data/2023/q3/data.csv
/path/to/sales-data/2023/q4/data.csv
/path/to/sales-data/2024/q1/data.csv
...
```

To import all `data.csv` files from all quarters across all years, you can use a wildcard for each directory level:

```sql
IMPORT INTO sales FROM '/path/to/sales-data/*/*/data.csv';

> ```

Author: Thanks for the clarification. I misread the https://github.com/pingcap/tidb/pull/59543 implementation, and I realized we still don't support that case. I'll revert the changes in this docs PR https://github.com/pingcap/docs/pull/21197 to focus on https://github.com/pingcap/tidb/pull/59543 changes and create a separate proposal for further enhancements.

> #### Import data from a nested directory structure
> 
> You can use wildcards to import data from a common directory structure, such as one organized by date. For example, assume your sales data is stored in S3 and organized by year and quarter:
> 
> ```
> /path/to/sales-data/2023/q1/data.csv
> /path/to/sales-data/2023/q2/data.csv
> /path/to/sales-data/2023/q3/data.csv
> /path/to/sales-data/2023/q4/data.csv
> /path/to/sales-data/2024/q1/data.csv
> ...
> ```
> 
> To import all `data.csv` files from all quarters across all years, you can use a wildcard for each directory level:
> 
> ```sql
> IMPORT INTO sales FROM '/path/to/sales-data/*/*/data.csv';
> ```
> 

Author: As clarified by @River2000i, the underlying code implementation still does not support directory-level wildcards. This whole new example isn't valid yet.

```suggestion
```
